### DISCUSSION TITLE
Allow structures to have a "homepage"

### DISCUSSION THREAD

**Discussion by leevigraham** [craftcms/cms#8390] [Ideas]
### Description

Let's say I have a section of my site called "Services" with urls that look like:

* /services
* /services/web-development
* /services/web-development/craft-cms
* /services/web-design

and I want to limit editing of these entries to a particular user / group.

There's two ways I could go about this:

Option 1. Create a new structure

I could create a new structure called "Services" with the `Entry URL Format` set to `{slug}`. A parent page would need to be created called `Services` with the `uri`: `/services`. This would list all the services and probably be a different entry type `Services Index`. Then authors could publish sub pages below 'Services' with a different entry type `Service Detail`.

The issue with this approach is that an author could publish an entry with any URL becuase they are not limited to just `/services` as defined by the 'Entry URL format' structure setting.

Option 2: Structure + Single

An alternative is to create a single for 'Services' with a locked down url and a structure for the Service details. This allows for fine grained permissions on both the index and the structure. The only downside is that the author needs to look in Singles to change the index, and 'Services' structure to change the entries. Using [Expanded Singles](https://github.com/engram-design/ExpandedSingles) makes this a bit easier but then there are two 'Services' links in the sidebar.

Option 3: Allow an entry to be the 'homepage' for a structure.

I'm thinking adding an option to structure entries for 'This is for the *structure* homepage' could work here. If the checkbox was clicked the uri for the structure would be set to `/` instead of the slug.

I'm already hacking something similar with a 'Site Structure' by setting the slug for one entry to `__home__`.

![image](https://cloud.githubusercontent.com/assets/25124/23821327/0bd76c20-0683-11e7-81ac-6de103ddbec5.png)

**Comment by leevigraham** [comment]


**Reply by brandonkelly** [reply]
Not sure I’m following?

**Comment by hiasl** [comment]
@leevigraham Setting the slug of a structure entry to `"__home__"` works for you? Is the screenshot above with the home symbol fake or real?
I tried the same and it does not work for me. I get URLs like `/baseUrlStructure/__home__` which is not my expected behavior :-)

**Reply by brandonkelly** [reply]
That would only work if the section’s URI format is just `{slug}`. Or at least something like:

```twig
{% if slug == '__home__' %}__home__{% else %} … normal URI format … {% endif %}
```

**Reply by hiasl** [reply]
Sorry, I'm not sure if I understand what you explain. I want (the same as @leevigraham I guess): The `__home__` should not appear appear in my URL created in the structure. As he wrote, the _homepage_ of the structure

Will this work without conditions in the URL definition? I guess not...

But with conditions, it must rather be something like:
`/baseUrlStructure/{% if slug != '__home__' %}{slug}{% endif %}`
to get rid of the `__home__`

Am I right? And if I am, wasn't the idea of `__home__` to have an easy solution for an "emtpy" slug? thanks!

**Reply by brandonkelly** [reply]
“Homepage” in this context is specifically referring to the site’s homepage.

**Reply by leevigraham** [reply]
@hiasl  This discussion was a blast from the past!

Currently I just create a root page for the structure and hang all the children off that. Thinking about this again… you could write a validation behaviour so that only one root element can exist.

This makes more sense visually as well given that if you could give an entry the slug `__home__` then the other pages will be displayed as siblings… but are children in a URL sense.

**Reply by hiasl** [reply]
@brandonkelly as you wrote, the URI format must be `{slug}` only, but `/sectionABC/{slug}` does not work. Tried it and can confirm this now. 

What is the reason for this? Would't it be ok to replace `__home__` with nothing when creating the URL and as long as no other URL is the same, it is fine?

**Comment by hiasl** [comment]
OK, then +1 for being able to set a homepage (slug =` __home__`) within structures, which is treated like an empty slug.

**Comment by thupsi** [comment]
I have shared these two somewhat relevant ideas: [#9736](https://github.com/craftcms/cms/discussions/9736), [#9783](https://github.com/craftcms/cms/discussions/9783).

In short, I'm asking for a way to relate an entry to a section or/and a way to assign fields to a section as a whole. This way a section (any section, not only structures) could even have **multiple homepages**! There are of course UI details that should be investigated to make this intuitive.
